odxtools/basicstructure.py

```diff
@@ -266,20 +258,21 @@ def encode(self, coded_request: Optional[bytes] = None, **params: ParameterValue
             is_end_of_pdu=True)
 
     def decode(self, message: bytes) -> ParameterValueDict:
-        # dummy decode state to be passed to convert_bytes_to_physical
         decode_state = DecodeState(coded_message=message)
-        param_values, cursor_byte_position = self.convert_bytes_to_physical(decode_state)
-        if not isinstance(param_values, dict):
-            odxraise(f"Decoding a structure must result in a dictionary of parameter "
-                     f"values (is {type(param_values)})")
-        if len(message) != cursor_byte_position:
+        param_values = self.decode_from_pdu(decode_state)
+
+        if len(message) != decode_state.cursor_byte_position:
             warnings.warn(
-                f"The message {message.hex()} is longer than could be parsed."
-                f" Expected {cursor_byte_position} but got {len(message)}.",
+                f"The message {message.hex()} probably could not be completely parsed:"
+                f" Expected length of {decode_state.cursor_byte_position} but got {len(message)}.",
                 DecodeError,
                 stacklevel=1,
             )
-        return param_values
+
+        if not isinstance(param_values, dict):
```

Comment:
With the semantic knowledge of what's inside `self.decode_state()`, it can't. That said, mypy is not yet able to solve the halting problem. so it does not look inside of function bodies and thus complains...